dev/gen_version.sh

```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+
+# This file should be called from the top-level directory of Coq!
+export LC_ALL=C
+COMPILE_DATE=`date`
+
+# default coq version
+COQ_VERSION=`cat COQ_VERSION || echo "unknown"`
+
+echo "type version = { describe : string; branch : string; hash : string } ;;"
+
+if [ -x `which git` ] && [ -d .git ]
+then
+    GIT_DESCRIBE=`git describe --tags --first-parent --dirty | sed 's/^V//'`
+    GIT_BRANCH=`git branch -a | sed -ne '/^\* /s/^\* \(.*\)/\1/p'`
```

Comment:
Two questions:
1. Why `git branch -a` rather than `git branch`.  I think `git` will always display the current branch, even without `-a`
2. Do we want to be using `git branch` at all here?  For example, `git symbolic-ref -q --short HEAD` gives a better differentiation between "we're on a branch" and "we're not on a branch" cases (how useful is the string `(no branch, rebasing (null))` from `git branch`?).  Here's a table with four different ways of doing things.

The four commands are:
1. `git symbolic-ref -q --short HEAD`
2. `git symbolic-ref -q HEAD`
3. `git branch | sed -ne '/^\* /s/^\* \(.*\)/\1/p'`
4. `git reflog | head -n 1 | sed -n s'/.*moving from .* to \(.*\).*/\1/p'`

| Command | `git checkout master` | `git checkout master^` | `git checkout origin/v8.0` | middle of `git rebase -i` | dangling commit* | fresh clone |
| --- | --- | --- | --- | --- | --- | --- |
| 1 | `master` | | | | | `master` |
| 2 | `refs/heads/master` | | | | | `refs/heads/master` |
| 3 | `master` | `(HEAD detached at 96a8d73)` | `(HEAD detached at upstream/v8.0)` | `(no branch, rebasing (null))` | `(HEAD detached from a4718ef)`** | `master` |
| 4 | `master` | `master^` | `upstream/v8.0` | | | |

*dangling commit = `git checkout a4718ef && touch foo && git add foo && git commit -m "tmp"`
**Note that the commit hash provided by `git branch` is misleading